_posts/2017-08-27-Clean-Architecture.md

```diff
@@ -0,0 +1,60 @@
+
+![Random built item](/img/lego-simple.jpg)
+
+## Professional lego building
+Now imagine if you got into this on a professional level. And you have these huge boxes of unsorted lego's. And they come to you with the question of building something quite large.
+
+![Lego cathedral](/img/cathedral.jpg)
+
+If you would need to start building a lego cathedral you'd probably want to spent a bit more time investing in structure. Structure in the way you work. Consistency in how you build certain parts. Making sure you don't loose time finding where pieces are. And making sure that when you need more help you can easily explain to people how they can help you and what they should do.
+
+This is the moment where things should start to become professional, stepping it up a notch, rather than just going along and clicking some randomly selected blocks together.
+
+![Sorted lego's](/img/sorted-lego.jpg)
+
+## Clean architecture
+The point of this article is not to explain what clean architecture is. Although currently it is definitly my preferred way of organizing my applications. There are already some great articles:
+
+* Uncle Bob's https://8thlight.com/blog/uncle-bob/2012/08/13/the-clean-architecture.html
+* Lieven Doclo's https://insaneprogramming.be/article/2017/02/14/thoughts-on-clean-architecture/
+
+Although Uncle Bob provides 2 drawings, his "circles in circles" is a bit confusing, so I tried combining his famous 2 images (cacircles & caparts) into a single overview.
+
+
+![Clean architecture](/img/clean-architecture.png)
+
+### A practical Java example
+
+TODO java callstack, rest -> app-api -> app-impl(ResolveTask) -> repo(get) -> domain(task.resolve()) -> repo(save)
+
+TODO color coded version with arrow to drawing
+
+### Why?
+##### Safety
+You usually don't work alone on your cathedral, you have senior developers, you have junior developers. CA is about having some boxes where they can put their parts. Not going into the whole title's in IT discussions but more experienced developers might have more problems putting their bricks into boxes. They've been throwing them into one big box, or 3 big boxes stacked on top of each other their whole carreer. Why should they change? Well often it's unprofessional and they don't mind sitting in a mess because they are used to it. But then when the complaints start rolling in about the code they are often already working on something new.
+
+Make it easy for everbody, also on yourself, even if you have plenty of experience. I just use one simple rule: "if can auto complete a class, you can use it." In order to ensure that, some structure will be necessary. I will not allow you to use jpa annotations on your rest controller. I will not allow you to call a repository from a controller. I will not allow you to call the postgres jdbc driver from the file upload servlet.
```

Comment:
akkoord met t sentiment maar zou ik wat ander verwoorden. Good design: making it easy to do something right, hard to do something wrong.... en van daaruit dan met de voorbeelden aankomen.